### ISSUE TITLE
Compiler warnings about MOOSE dummies

### ISSUE THREAD

**Issue by roystgnr** [idaholab/moose#13914]
## Bug Description
Every Registry.h macro invocation generates an 11 line long compiler warning for me, e.g.:

```
/home/roystgnr/git/moose-fresh/framework/build/header_symlinks/Registry.h:47:28: warning: ‘dummyvar_for_registering_obj_SmoothMeshGenerator19’ defined but not used [-Wunused-variable]
   static char combineNames(dummyvar_for_registering_obj_##classname, __LINE__) =                   \
                            ^
/home/roystgnr/git/moose-fresh/framework/build/header_symlinks/Registry.h:17:29: note: in definition of macro ‘combineNames1’
 #define combineNames1(X, Y) X##Y
                             ^
/home/roystgnr/git/moose-fresh/framework/build/header_symlinks/Registry.h:47:15: note: in expansion of macro ‘combineNames’
   static char combineNames(dummyvar_for_registering_obj_##classname, __LINE__) =                   \
               ^
/home/roystgnr/git/moose-fresh/framework/src/meshgenerators/SmoothMeshGenerator.C:19:1: note: in expansion of macro ‘registerMooseObject’
 registerMooseObject("MooseApp", SmoothMeshGenerator);
```

## Steps to Reproduce
Build MOOSE with the -Wunused-variable compiler flag or with one of the many flags that incorporates it.

## Impact
This is just an annoyance, but it's a voluminous one: tens of thousands of lines of warning text in a typical MOOSE build.

What's the point of the static char here?

**Comment by permcody** [comment]
@rwcarlsen can expand here but he's creating a static function that can be called during the static initialization phase. This is all tricky business since the C++ runtime doesn't have the most well-defined static initialization behavior. He said he has to assign the function value to a variable for it to be a valid statement.

I swore that we used that compiler flag in at least one of our build configs, but perhaps I'm mistaken? Can we just wrap the thing in a pragma ignore?

**Comment by roystgnr** [comment]
> He said he has to assign the function value to a variable for it to be a valid statement.

Ah, of course, I should have figured that out.

> Can we just wrap the thing in a pragma ignore?

That'd probably do it.  But now you've got me wondering why I never saw the same warnings when using the libMesh factory pattern in apps.  IIRC that works the same way, with global objects that do their real work in their constructors and then never get used again.  Is the compiler just less picky about unused variables if they're in classes than POD?

**Comment by permcody** [comment]
Roy are you still seeing this error? If so, I'd be interested in hearing about how to repeat. We don't see it with new compilers on our linux systems. Is it some set of flags you use in libMesh?

**Comment by roystgnr** [comment]
Huh.  Now that you mention it, I didn't notice any of this in my last MOOSE build.  That might have been on a different system, though.  Let me try again and see if I can still replicate it.

**Comment by roystgnr** [comment]
And I can't.  Not on multiple systems, at least.  I'm an idiot for not recording exactly what compiler and settings and so on I was using in the first place, so I can't say for sure whether this was somehow fixed or whether I only triggered it on a particularly esoteric build, but if it's not a problem we can trigger then for now at least it's not a problem.
